TypeCobol/Compiler/Sql/Scanner/SqlScanner.cs

```diff
@@ -50,46 +80,242 @@ public override Token GetTokenStartingFrom(int startIndex)
+                case 'g':
+                    if ((startIndex + 2 <= lastIndex) && (line[startIndex + 1] == 'x' || line[startIndex + 1] == 'X') &&
+                        (line[startIndex + 2] == '\'' || line[startIndex + 2] == '"'))
+                    {
+                        return ScanAlphanumericLiteral(startIndex, TokenType.SQL_GraphicStringLiteral, null);
+                    }
+                    else
+                    {
+                        return ScanUserDefinedWord(startIndex);
+                    }
+                default:
+                    return ScanUserDefinedWord(startIndex);
 
             }
+        }
 
-            if (IsSqlKeywordPart(line[currentIndex]))
+        private Token ScanSqlNumericLiteral(int startIndex, bool hasExplicitSign)
+        {
+            if (TryScanDecimalFloatingPointSpecialValue(startIndex, hasExplicitSign, out var decimalFloatingPointLiteralTokenValue))
             {
-                //Consume all sql-keyword compatible chars
-                for (; currentIndex <= lastIndex && IsSqlKeywordPart(line[currentIndex]); currentIndex++){}
-                string tokenText = line.Substring(startIndex, currentIndex - startIndex);
-                //Try to match keyword text
-                var tokenType = TokenUtils.GetSqlKeywordTokenTypeFromTokenString(tokenText);
-                switch (tokenType)
+                Token decimalFloatingPointSpecialValueToken = new Token(TokenType.SQL_DecimalFloatingPointLiteral,
+                    startIndex, currentIndex - 1, tokensLine)
+                {
+                    LiteralValue = decimalFloatingPointLiteralTokenValue
+                };
+                return decimalFloatingPointSpecialValueToken;
+            }
+
+            var token = ScanNumericLiteral(startIndex, '.');
+            switch (token.TokenType)
+            {
+                case TokenType.IntegerLiteral:
+                    CheckIntegerLiteral();
+                    break;
+                case TokenType.DecimalLiteral:
+                    CheckDecimalLiteral();
+                    break;
+                case TokenType.FloatingPointLiteral:
+                    CheckFloatingPointLiteral();
+                    break;
+            }
+
+            return token;
+
+            void CheckIntegerLiteral()
+            {
+                Debug.Assert(token.LiteralValue is IntegerLiteralTokenValue);
+                var literalValue = (IntegerLiteralTokenValue)token.LiteralValue;
+                int literalValueLength = literalValue.Number.ToString().Length -
+                                         (literalValue.HasSign == true ? 1 : 0);
+                //Check IntegerLiteral range
+                if (literalValueLength > 19)
+                {
+                    //Out of range
+                    this.tokensLine.AddDiagnostic(MessageCode.SyntaxErrorInParser, token,
+                        "Number is too big : " + literalValue.Number);
+                }
```

Comment:
From DB2 spec:

> A decimal constant is a signed or unsigned number of no more than 31 digits and either includes a decimal point or is not within the range of binary integers.

So this means that we should not add a diagnostic here but correct the type and the value to switch to a DecimalLiteral and then check the DecimalLiteral.